### ISSUE TITLE
Error in getUserInfo Error: parseAndCheckLogin got status code: 404

### ISSUE THREAD

**Issue by cvigoe** [Schmavery/facebook-chat-api#229]
Getting a peculiar issue where I get an error after successfully logging in and successfully getting thread list. I am trying to convert the user IDs in the getThreadList using getUserInfo and this is where all the errors arise. Interestingly enough, the errors are only showing up when I have deployed the app on Heroku, I hadn't encountered any of these sort of problems before when serving locally. Not sure if this is an issue with the api or something else, but I think it is related to an issue with getUserInfo, would greatly appreciate any feedback on this.

<img width="1280" alt="screen shot 2016-03-31 at 11 23 36 am" src="https://cloud.githubusercontent.com/assets/6818920/14186355/d0fceee6-f732-11e5-80c1-5fc859131115.png">

**Comment by ElTimuro** [comment]
+1

**Comment by Schmavery** [comment]
As with any issue that involves using the results of one function to call another, have you checked what is returned from the first call?  Can we see what you're actually calling getUserInfo with?

**Comment by Schmavery** [comment]
@ElTimuro Can you elaborate on your +1?

**Comment by ElTimuro** [comment]
Just wanted to say me too..

**Comment by ElTimuro** [comment]
It's strange, because the same code worked before. Didn't change it for Weeks..

**Comment by Schmavery** [comment]
@cvigoe said it worked for him locally but not on heroku, is this the same for you?
It's reasonably normal that stuff breaks on its own now and then because of the nature of this api relying on facebook.  It looks like your code is breaking in sendMessage...  Is this error message wrong or are you hitting this error in a different situation?  If you can post the smallest self-contained example you can that reproduces this behaviour, that might be helpful.

**Comment by ElTimuro** [comment]
I was successfully hosting it on Uberspace (linux based share server).

`var login = require("facebook-chat-api");
var request = require('request-json');
var fs = require('fs');
var https = require('https');


login({ email: "********", password: "******" }, function callback(err, api) {
    if (err) return console.error(err);

    api.listen(function callback(err, message) {
        var m = message.body.replace(/\?/g, '');
        m = m.replace(/\./g, '');

        if (!m) {
            m = 'emptymessage';
        }

        var path = "/api/facebook/" + message.senderID + "/" + encodeURIComponent(m);
        console.log(path);

        https.get({
            host: '*******',
            path: path,
            port: 443
        }, function(response) {
            // Continuously update stream with data
            var body = '';
            response.on('data', function(d) {
                body += d;
            });
            response.on('end', function() {
                var response = body.substring(1, body.length - 1);
                api.sendMessage(response, message.threadID);
            });
        });
    });
});`

**Comment by ElTimuro** [comment]
Yes looks like something changed on the fb end. message in api.listen(function callback(err, message) was undefined, last time i checked..

**Comment by cvigoe** [comment]
@Schmavery 
What initially appeared to be 1 problem now seems to be 2:

1: Some accounts when attempting to log in with the facebook-chat-api result in an "Error retrieving userID" error. Sometimes going to facebook in the browser and completing the security check and then attempting to log in using the facebook-chat-api again clears this issue, sometimes the same issue appears if not immediately, at some point again in the future.

2: Some accounts successfully log in with the api but when getUserInfo is called using the api object, a separate error to the error mentioned above occurs: "Error in getUserInfo Error: parseAndCheckLogin got status code: 404".

Both errors seem to be pretty unpredictable, and I can't really figure out what's causing them. Facebook seems to sometimes lock the user's account when the Heroku server uses the api as a security measure, but sometimes it doesn't. I don't know if this is something that the api causes or if it's something to do with the Heroku server, I suspect it's the latter as when I log in to facebook to verify my account in the browser, it asks me to confirm if the recent login attempt was by me and lists the Heroku server's location, which obviously isn't my usual location. I suspect this has something to do with the Heroku server as I can't recall encountering this type of error until I hosted it on Heroku. The second error I can't explain with facebook's security, the api successfully logs in and even manages to getThreadList, but it encounters a strange 404 parse error when attempting to make another api call with getUserInfo. I'd appreciate any help on this / advice on a workaround of some sort with either of these problems!

**Comment by Schmavery** [comment]
The issue with heroku causing facebook to lock the account occasionally is somewhat of a known issue that we've experienced in our own bots that use heroku.  This may be due to heroku spinning up on your process on IPs that look suspicious to Facebook.  There is no known workaround to this, though attempting to use appState may help.

As I said before, I'd like to attempt to establish whether your second problem is happening in getThreadList or getUserInfo before any further investigation there.

**Comment by cvigoe** [comment]
I've been testing locally and I've been able to eliminate the random nature of this error, but I still can't figure out how to fix it. When I run my server locally and use the api to log in to my facebook account, everything works fine. I can log out successfully too. I can also log in and out using the same account as many times as I like without any issues. As soon as I try to log in with a different facebook account to the first one, I start encountering this "Error in getUserInfo" problem. After attempting to log in with this second account, I also encounter the same problem when I go back and try and log in using the first account that originally posed no problems. If I restart the server and try logging in with the second account (the one that caused this 404 error) it successfully logs in and out. When I try logging in with the first account (the one that was fine originally) I get the same 404 issue. So obviously whatever account that first logs in after the server starts up, will continue to be ok until a new account is tried, at which point all future attempts to log in using the api will fail with the same error. The code for the web app I'm developing is here [https://github.com/cvigoe/href](url).

Again, any feedback would be appreciated - thanks to @Schmavery for getting back so promptly, solid support!

**Comment by Schmavery** [comment]
> As soon as I try to log in with a different facebook account to the first one

I don't understand, sorry...  Suddenly there are multiple accounts involved in this issue?

**Comment by cvigoe** [comment]
@Schmavery if you take a look at href.herokuapp.com you might get a better idea at what I'm trying to do

**Comment by Schmavery** [comment]
If the problem is with your app, unfortunately you're probably the best person to find the bug.  Otherwise if you suspect the problem is with the api, it will be most useful for us if you are able to isolate the issue to a small example to reproduces the problem.  This will save us a lot of time trying to understand and run your code.

**Comment by bestkao** [comment]
@cvigoe you mention logging in with another facebook account. Is this via the web browser? If so, I'm running across a similar problem here #236

I'm thinking that having an active browser session while using the api locally starts creating issues. I haven't tried it on production though. Maybe Facebook isn't too fond of conflicting logins on the same IP address..

**Comment by cvigoe** [comment]
@Schmavery @bestkao Ok I figured out what was causing the problems - after logging in with the api, I got the user's thread history and then attempted to get the names of the participants of the thread using getUserInfo by passing an array of user IDs that I populated by simply iterating through the thread history object and adding each participant ID I came across into an array. When I logged out, I wasn't resetting this array, so when I tried logging in with a separate account, the login would be successful but when I attempted to get the names of the participants of the threads, my app was attempting to get the names of users that were in the thread of the previous login, hence a 404. Thanks for the active support  @Schmavery , and I'm posting this in case anyone else seems to be encountering similar issues.
